### PR TITLE
Refactoring component_base and base_action/transfer_base_action 

### PR DESCRIPTION
**Pull request by hkaiser**
This allows to reduce the number of instantiated functions and exported symbols, especially for external modules that contain components.

### CODE REVIEW

File: src/runtime/components/server/component_base.cpp [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        // returned at this point will control the lifetime of the
+        // component.
+        naming::detail::set_credit_split_mask_for_gid(gid);
+        return gid;
+    }
+
+    naming::address component_base::get_current_address() const
+    {
+        return naming::address(get_locality(), type_,
+            std::uint64_t(static_cast<component_base const*>(this)));
```
Comment (sithhell) [inline src/runtime/components/server/component_base.cpp]: the static cast here is not necessary, is it?

File: hpx/runtime/components/component_type.hpp [hunk 2]
```diff
@@ -174,12 +174,12 @@ namespace hpx { namespace components
 
     // Returns the (unique) name for a given component
     template <typename Component, typename Enable = void>
-    const char* get_component_name();
+    HPX_CONSTEXPR char const* const get_component_name();
```
Comment (sithhell) [inline hpx/runtime/components/component_type.hpp]: the extra const generates warnings for clang.

File: hpx/runtime/components/component_type.hpp [hunk 3]
```diff
@@ -174,12 +174,12 @@ namespace hpx { namespace components
...
 
     // Returns the (unique) name of the base component. If there is none,
     // nullptr is returned
     template <typename Component, typename Enable = void>
-    const char* get_component_base_name();
+    HPX_CONSTEXPR const char* const get_component_base_name();
```
Comment (sithhell) [inline hpx/runtime/components/component_type.hpp]: same here.

File: hpx/runtime/components/component_type.hpp [hunk 4]
```diff
@@ -253,13 +253,13 @@ namespace hpx { namespace components
 
 #define HPX_DEFINE_COMPONENT_NAME_2(Component, name)                          \
 namespace hpx { namespace components {                                        \
-    template <> HPX_ALWAYS_EXPORT                                             \
-    const char* get_component_name< Component, void>()                        \
+    template <> HPX_CONSTEXPR                                                 \
+    char const* const get_component_name< Component, void>()                  \
```
Comment (sithhell) [inline hpx/runtime/components/component_type.hpp]: same here.

File: hpx/runtime/components/component_type.hpp [hunk 5]
```diff
@@ -253,13 +253,13 @@ namespace hpx { namespace components
...
     {                                                                         \
         return HPX_PP_STRINGIZE(name);                                        \
     }                                                                         \
-    template <> HPX_ALWAYS_EXPORT                                             \
-    const char* get_component_base_name< Component, void>()                   \
+    template <> HPX_CONSTEXPR                                                 \
+    char const* const get_component_base_name< Component, void>()             \
```
Comment (sithhell) [inline hpx/runtime/components/component_type.hpp]: same here.

File: hpx/runtime/components/component_type.hpp [hunk 6]
```diff
@@ -268,13 +268,13 @@ namespace hpx { namespace components {                                        \
 
 #define HPX_DEFINE_COMPONENT_NAME_3(Component, name, base_name)               \
 namespace hpx { namespace components {                                        \
-    template <> HPX_ALWAYS_EXPORT                                             \
-    const char* get_component_name< Component, void>()                        \
+    template <> HPX_CONSTEXPR                                                 \
+    char const* const get_component_name< Component, void>()                  \
```
Comment (sithhell) [inline hpx/runtime/components/component_type.hpp]: same here.

File: hpx/runtime/components/component_type.hpp [hunk 7]
```diff
@@ -268,13 +268,13 @@ namespace hpx { namespace components {                                        \
...
     {                                                                         \
         return HPX_PP_STRINGIZE(name);                                        \
     }                                                                         \
-    template <> HPX_ALWAYS_EXPORT                                             \
-    const char* get_component_base_name< Component, void>()                   \
+    template <> HPX_CONSTEXPR                                                 \
+    char const* const get_component_base_name< Component, void>()             \
```
Comment (sithhell) [inline hpx/runtime/components/component_type.hpp]: same here.

File: src/runtime/components/server/component_base.cpp (hunk 1 above)
Comment (hkaiser) [reply src/runtime/components/server/component_base.cpp]: It is needed as `this` might be different from the base address of the whole component.

Comment (hkaiser) [reply src/runtime/components/server/component_base.cpp]: But this code is wrong for other reasons, will fix.

**Comment by sithhell** [comment]
appveyor is not happy with this right now

**Comment by hkaiser** [comment]
> appveyor is not happy with this right now

Yes, I know - but this was introduced earlier by one of the changes to the thread scheduling.

**Comment by sithhell** [comment]
> Yes, I know - but this was introduced earlier by one of the changes to the thread scheduling.

Can you restart the build then? It looks like this PR is the only one affected to that timeout.

**Comment by hkaiser** [comment]
The same happened here: https://ci.appveyor.com/project/hkaiser/hpx/build/fix_timed_suspension-1897, apparently this got branched off at a bad time... I'll rebase and restart the tests.

**Pull request merged**